src/Microsoft.NetCore.Analyzers/Core/Security/DoNotUseInsecureSettingsForJsonNet.cs

```diff
@@ -0,0 +1,260 @@
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.ValueContentAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.NetCore.Analyzers.Security.Helpers;
+
+namespace Microsoft.NetCore.Analyzers.Security
+{
+    /// <summary>
+    /// For detecting potentially insecure deserialization settings with <see cref="T:Newtonsoft.Json.JsonSerializerSettings"/>.
+    /// </summary>
+    [SuppressMessage("Documentation", "CA1200:Avoid using cref tags with a prefix", Justification = "The comment references a type that is not referenced by this compilation.")]
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    class DoNotUseInsecureSettingsForJsonNet : DiagnosticAnalyzer
+    {
+        // TODO paulming: Help links URLs.
+        internal static readonly DiagnosticDescriptor DefinitelyInsecureSettings =
+            SecurityHelpers.CreateDiagnosticDescriptor(
+                "CA2327",
+                nameof(MicrosoftNetCoreSecurityResources.JsonNetInsecureSettingsTitle),
+                nameof(MicrosoftNetCoreSecurityResources.JsonNetInsecureSettingsMessage),
+                isEnabledByDefault: false,
+                helpLinkUri: null);
+        internal static readonly DiagnosticDescriptor MaybeInsecureSettings =
+            SecurityHelpers.CreateDiagnosticDescriptor(
+                "CA2328",
+                nameof(MicrosoftNetCoreSecurityResources.JsonNetMaybeInsecureSettingsTitle),
+                nameof(MicrosoftNetCoreSecurityResources.JsonNetMaybeInsecureSettingsMessage),
+                isEnabledByDefault: false,
+                helpLinkUri: null);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(
+                DefinitelyInsecureSettings,
+                MaybeInsecureSettings);
+
+        /// <summary>
+        /// PropertySetAbstractValue index for TypeNameHandling property.
+        /// </summary>
+        private const int TypeNameHandlingIndex = 0;
+
+        /// <summary>
+        /// PropertySetAbstractValue index for Binder / SerializationBinder properties (both are aliased to same underlying value).
+        /// </summary>
+        private const int SerializationBinderIndex = 1;
+
+        private static readonly ConstructorMapper ConstructorMapper = new ConstructorMapper(
+            (IMethodSymbol constructorMethod, IReadOnlyList<PointsToAbstractValue> argumentPointsToAbstractValues) =>
+            {
+                if (constructorMethod.Parameters.Length == 0)
+                {
+                    return PropertySetAbstractValue.GetInstance(
+                        PropertySetAbstractValueKind.Unflagged,   // TypeNameHandling defaults to None.
+                        PropertySetAbstractValueKind.Flagged);    // Binder / SerializationBinder defaults to null.
```

Comment:
Why there are only 2 kinds?

Reply:
The Binder and SerializationBinder properties are both effectively the same value.

---
In reply to: [296158716](https://github.com/dotnet/roslyn-analyzers/pull/2415#discussion_r296158716) [](ancestors = 296158716)